src/databricks/labs/ucx/assessment/crawlers.py

```diff
@@ -250,3 +269,150 @@ def snapshot(self) -> list[ClusterInfo]:
+            )
+        )
+        all_jobs = list(self._ws.jobs.list())
+        all_clusters: dict[str, ClusterDetails] = {c.cluster_id: c for c in self._ws.clusters.list()}
+        return self._assess_job_runs(all_clusters, all_job_runs, all_jobs)
+
+    def _create_hash_from_job_run_task(self, task: RunTask) -> str:
+        """
+        Check all hashable attributes, append to a singular list, remove None's,
+        pipe separate and hash the value to provide a uniqueness identifier
+        - specifically ignore parameters as these change.
+        """
+        hash_values = []
+        if task.notebook_task is not None:
+            hash_values.append(task.notebook_task.notebook_path)
+        if task.spark_python_task is not None:
+            hash_values.append(task.spark_python_task.python_file)
+        if task.spark_submit_task is not None:
+            hash_values.append(task.spark_submit_task.parameters)
+        if task.spark_jar_task is not None:
+            hash_values.append(task.spark_jar_task.jar_uri)
+            hash_values.append(task.spark_jar_task.main_class_name)
+        if task.pipeline_task is not None:
+            hash_values.append(task.pipeline_task.pipeline_id)
+        if task.python_wheel_task is not None:
+            hash_values.append(task.python_wheel_task.package_name)
+            hash_values.append(task.python_wheel_task.entry_point)
+        if task.sql_task is not None:
+            hash_values.append(task.sql_task.file.path)
+            hash_values.append(task.sql_task.alert.alert_id)
+            hash_values.append(task.sql_task.dashboard.dashboard_id)
+            hash_values.append(task.sql_task.query.query_id)
+        if task.dbt_task is not None:
+            task.dbt_task.commands.sort()
+            hash_values.append(task.dbt_task.schema)
+            hash_values.append(task.dbt_task.catalog)
+            hash_values.append(task.dbt_task.warehouse_id)
+            hash_values.append(task.dbt_task.project_directory)
+            hash_values.append(",".join(task.dbt_task.commands))
+        if task.condition_task is not None:
+            hash_values.append(task.condition_task.op.value)
+            hash_values.append(task.condition_task.right)
+            hash_values.append(task.condition_task.left)
+            hash_values.append(task.condition_task.outcome)
+        if task.run_job_task is not None:
+            hash_values.append(task.run_job_task.job_id)
+        if task.git_source is not None:
+            hash_values.append(task.git_source.git_url)
+            hash_values.append(task.git_source.git_tag)
+            hash_values.append(task.git_source.git_branch)
+            hash_values.append(task.git_source.git_commit)
+            hash_values.append(task.git_source.git_provider)
+            hash_values.append(task.git_source.git_snapshot.used_commit)
+
+        hash_value_string = "|".join([value for value in hash_values if value is not None])
```

Comment:
```suggestion
        hash_value_string = "|".join([str(value) for value in hash_values if value is not None])
```
as some of the things may not be strings. python is a language that allows sloppy code and we have to be very defensive against the language itself. probably surround it with `try .. except` to be extra careful :D

Reply:
Shouldn't the SDK be enforcing typing?